Jetpack E2E: add flow for AI Assistant block.

worldomonation opened the pull request:
Related to https://github.com/Automattic/wp-calypso/issues/80730.

## Proposed Changes

This PR adds a new block testing flow for the AI Assistant.

Key flows:
- initial query
- change tone
- improve

This change stretches what we have done with the Block Smoke Testing/BlockFlow template. 

## Testing Instructions
Ensure the following build configurations are passing:
  - [x] Jetpack E2E Simple (desktop)
  - [x] Jetpack E2E Simple (mobile)
  - [x] Jetpack E2E AT 
  - [x] Calypso E2E (desktop)

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
packages/calypso-e2e/src/lib/blocks/block-flows/index.ts
- add new block testing flow.

test/e2e/specs/blocks/blocks__jetpack-earn.ts
- add new test case.

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- add some guards

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- add some more guards

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- additional waits
- fix JSDoc length.

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- refactor methods to work with both block-based and non-block-based themes.
- add waits throughout the methods

worldomonation committed:
Move tests to a new file `jetpack writing`

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- newline

dpasque commented on test/e2e/specs/blocks/blocks__jetpack-writing.ts:
```diff
@@ -0,0 +1,16 @@
+
+const blockFlows: BlockFlow[] = [
+	new AIAssistantFlow( {
+		query: 'In two short sentences, tell me about Vancouver, Canada.',
```
❤️

dpasque commented on packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts:
```diff
@@ -0,0 +1,158 @@
+	 * @param {Locator} block Locator to the block.
+	 */
+	private async waitForQuery( block: Locator ) {
+		await Promise.all( [
+			block
+				.getByRole( 'button', { name: 'Stop request' } )
+				.waitFor( { state: 'detached', timeout: 15 * 1000 } ),
+			block
+				.getByRole( 'textbox', { name: 'Ask Jetpack AI', disabled: true } )
+				.waitFor( { state: 'detached', timeout: 15 * 1000 } ),
+			block.locator( '.components-spinner' ).waitFor( { state: 'detached' } ),
+		] );
```
**Question (non-blocking):** Out of curiosity, why all three? These all seem to happen at once from everything I can tell, so I think we should be able to just key off of one, right? I think that would decrease the surface area of potential breaking DOM changes/flakiness, etc. WDYT?

worldomonation replied:
Good and valid question.

During development I consistently ran into issues with the test running under Jetpack AT environment racing past the waits. 

The first wait `Stop request` can probably be removed and instead we can rely on either the second or third waits.

dpasque replied:
Ah sure, it sounds like maybe they don't always reliably end at the same time! Thanks for explaining! 😄

dpasque commented on packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts:
```diff
@@ -0,0 +1,158 @@
+	 *
+	 * @param {Locator} block Locator to the block.
+	 */
+	private async waitForQuery( block: Locator ) {
```
**Nit (optional):** Tiny signature nit: this `Locator` almost always seems to be the editor canvas, not the block, so maybe we should make the parameter name and JSDOC comment reflect that?

dpasque commented on packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts:
```diff
@@ -0,0 +1,158 @@
+		const elementHandle = await toolbar.elementHandle();
+		await elementHandle?.waitForElementState( 'stable' );
+
+		const button = toolbar.getByRole( 'button', { name: buttonName } );
+		await button.waitFor();
+		await button.click();
+
+		const option = editorParent.getByRole( 'menuitem', { name: optionName } );
+		await option.waitFor();
+		await option.click();
```
**Suggestion (non-blocking):** We have component POM classes for these already! ([Block toolbar](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/components/editor-block-toolbar-component.ts) and [popover menu](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/components/editor-popover-menu-component.ts).) What do you think about extending/enhancing and using those, to keep things DRYish?

dpasque commented on packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts:
```diff
@@ -0,0 +1,158 @@
+		// Yes, yuck, I know. But Playwright often tries to act on the buttons sooner
+		// than it should, so we need to wait for the bounding box of the element here
+		// to be stable before attempting to click on the toolbar.
+		const elementHandle = await toolbar.elementHandle();
```
**Question (non-blocking):** Out of curiosity, can you explain what was going on here? I thought playwright auto-waited for `stable` on `click` actions, so I'm wondering if there's something else going on? 

https://playwright.dev/docs/actionability

worldomonation replied:
>  I thought playwright auto-waited for stable on click actions, so I'm wondering if there's something else going on?

Yes, Playwright does wait - and on the local machine, the spec doesn't need this little yucky part in order to pass _most of the time_. 

However, sometimes on local machine, and more frequently on CI, Playwright's auto-actionability checks simply race past the face that the Editor Toolbar isn't yet actually fully ready. 

As far as I can tell, the following happens in this scenario:

1. Playwright begins waiting for a button on the Editor Toolbar. AI Assistant block is generating and filling in text.
2. Once the AI Assistant block is done filling in text, the button begins to appear on the Editor Toolbar. 
3. Playwright resolves the waiting part, and begins actionability checks on the element. This, for some reason, passes when it should not, because the Editor Toolbar is still in the process of "expanding" in a sense, by rendering all the available action buttons.
4. Playwright thinks the target button is ready to click, so it performs the action. Meanwhile, the Editor Toolbar is not yet in fact ready, but the action gets performed on it anyway. 
5. The Editor Toolbar ends up "swallowing" the action, and the expected Popover doesn't show.
6. Playwright fails to locate the option in the Popover because it isn't there.

worldomonation replied:
Case in point, I was able to get this portion to fail relatively consistently on local machine in Jetpack AT environment, and it stopped as soon as I put the `waitForElementState( stable )` call.

To me, it seems the Playwright's actionability waits only apply to the immediate locator you're trying to action upon. For example, in this case, the actionability checks determine the "Change tone" button is present and it's ready to interact with, but in reality the parent component "Editor Toolbar" is not yet stable.

worldomonation replied:
In the most recent revision I removed the use of ElementHandle and it seems to still do fine, so let's stick with that for now.

dpasque replied:
Ah sure, thank you so much for explaining all that! That makes sense, and is a bummer to have to work with! 😕 There could also be a React race condition here too where the element is in the DOM, but the handlers attached to it aren't included in the first render pass or something like that. Thanks for explaining! If it's running stable, I think we're good to go 👍  😄

dpasque reviewed (approved):
Had a few questions, but none blocking! Thanks for getting this script done, it's looking great!

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- refactor block flow to use already-existing methods.

test/e2e/specs/blocks/blocks__jetpack-writing.ts
- update spec

packages/calypso-e2e/src/lib/pages/editor-page.ts
packages/calypso-e2e/src/lib/components/editor-popover-menu-component.ts
packages/calypso-e2e/src/lib/components/editor-block-toolbar-component.ts
- expose method from the subcomponents into EditorPage.

worldomonation committed:
packages/calypso-e2e/src/lib/blocks/block-flows/ai-assistant.ts
- fix typo

worldomonation committed:
Remove use of trial

worldomonation committed:
packages/calypso-e2e/src/lib/components/editor-popover-menu-component.ts
- remove narrowing of selector

worldomonation commented:
I suppose you approved this already @dpasque, so no need to do another pass if you don't think it's warranted.

dpasque commented:
@worldomonation 

> I suppose you approved this already @dpasque, so no need to do another pass if you don't think it's warranted.

Yep! But I took a quick look anyway and it's looking really nice too -- thanks for making all the tweaks! 🚀
